object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -455,40 +475,52 @@ private void handleToBuilderType(
    */
   private List<String> getAutoValueRequiredProperties(
       final Element autoValueClassElement, Set<String> allBuilderMethodNames) {
-    List<String> requiredPropertyNames = new ArrayList<>();
-    for (Element member : autoValueClassElement.getEnclosedElements()) {
-      if (member.getKind().equals(ElementKind.METHOD)) {
-        // should be an abstract instance method
-        Set<Modifier> modifiers = member.getModifiers();
-        if (!modifiers.contains(Modifier.STATIC) && modifiers.contains(Modifier.ABSTRACT)) {
-          String name = member.getSimpleName().toString();
-          TypeMirror returnType = ((ExecutableElement) member).getReturnType();
-          if (!IGNORED_METHOD_NAMES.contains(name) && !returnType.getKind().equals(TypeKind.VOID)) {
-            // shouldn't have a nullable return
-            boolean hasNullable =
-                Stream.concat(
-                        elements.getAllAnnotationMirrors(member).stream(),
-                        returnType.getAnnotationMirrors().stream())
-                    .anyMatch(anm -> AnnotationUtils.annotationName(anm).endsWith(".Nullable"));
-            if (hasNullable) {
-              continue;
-            }
-            // if return type of foo() is a Guava Immutable type, not required if there is a builder
-            // method fooBuilder()
-            if (isGuavaImmutableType(returnType)
-                && allBuilderMethodNames.contains(name + "Builder")) {
-              continue;
-            }
-            // if it's an Optional, the Builder will automatically initialize it
-            if (isOptional(returnType)) {
-              continue;
-            }
-            requiredPropertyNames.add(name);
-          }
-        }
-      }
-    }
-    return requiredPropertyNames;
+    return getAllSupertypes((Symbol) autoValueClassElement).stream()
+        .flatMap(e -> e.getEnclosedElements().stream())
+        .filter(
+            member -> {
```

Comment:
This lambda is even more difficult to read - it's nearly a full page of code. I also think this should be its own method, with documentation describing the property it is computing.